src/upload.rs

```diff
@@ -0,0 +1,112 @@
+use std::collections::BTreeMap;
+use std::sync::{Arc, Mutex, Weak};
+
+use bytes::Bytes;
+use rand::distributions::Alphanumeric;
+use rand::{Rng, SeedableRng};
+use rand_chacha::ChaCha20Rng;
+use tracing::warn;
+
+pub struct UploadServer {
+    rng: ChaCha20Rng,
+    job_cache: BTreeMap<String, Weak<Mutex<JobArtifacts>>>,
+    base_addr: Option<String>,
+}
+
+impl UploadServer {
+    pub fn new() -> Self {
+        Self {
+            rng: ChaCha20Rng::from_entropy(),
+            job_cache: Default::default(),
+            base_addr: None,
+        }
+    }
+
+    pub fn add_new_job(&mut self) -> Arc<Mutex<JobArtifacts>> {
+        // Wipe any dead jobs as the new one starts. It's not hugely
+        // important when this happens, so long as it happens
+        // periodically.
+        self.cleanup();
```

Comment:
why is a periodic cleanup required? or rather cant the registration be dropped whenever the relevant jobs get dropped?.. Also I'm not a big fan of scatering `Arc<Mutex<T>>` around as it tends to make things harder to use from the user point of view :)